### DISCUSSION TITLE
[nuxt2 bridge/ nuxt3] ohmyfetch interceptor plugin

### DISCUSSION THREAD

**Discussion by psycongaroo** [nuxt/nuxt#16958] [Questions]
Hi guys,
I saw ohmyfetch has been updated and now interceptors are avaliable.
How can I create a plugin in nuxt2 bridge/nuxt 3 for create a shared fetch with interceptors?
I need that useFetch or useAsyncData will use interceptors too.

https://github.com/unjs/ohmyfetch/issues/19

Thank you so much!

**Comment by misaon** [comment]
Simple, in your plugin you can use global `$fetch` variable everywhere 👍🏼

**Reply by psycongaroo** [reply]
> Simple, in your plugin you can use global `$fetch` variable everywhere 👍🏼

Great idea, thank you so much.. 
but I need that useFetch or useAsyncData will use interceptors too.

**Reply by misaon** [reply]
You can use global `useFetch` and `useAsyncData` too.

**Reply by misaon** [reply]
My working implementation looks like this
![image](https://user-images.githubusercontent.com/15806356/147107766-00138118-24a7-4538-af74-6be24faaf005.png)

**Reply by pi0** [reply]
Small tip. You can use `$fetch.create({ ... })` to share common options :)

**Reply by psycongaroo** [reply]
Thank you both guys!

I mean is possible to create a nuxt plugin where I create a $fetch instance in it with interceptor onRequest onRequestError and pass automatically in original composables useFetch or useAsyncData?

So when we'll use those composables the interceptors are already injected?
@pi0

**Reply by misaon** [reply]
I think its not possible now.

**Reply by Sin1tar** [reply]
> 

![image](https://user-images.githubusercontent.com/82471918/147136581-25742bd7-7ab9-45c0-af86-90acc35388d5.png)

You can!
Your $fetch is now being processed from the plugin entry point.
However, the current solution will increase your entry.js by 7 kilobytes.
And I'm not sure if the current solution on the screenshot is rational.

**Reply by pi0** [reply]
@SergeyMustafin Replacing globalThis implementation is probably not the best idea. It causes SSR context leaks and might be readonly getter in the future versions of nuxt.   We were using this global pattern with `$axios` and `$http` but it can make lots of conflicts when another library is depending `$fetch` and does expect different defaults or handlers.

Instead we can simply create an auto imported composable in `composables/api.ts`:

```js
export const apiFetch = $fetch.create({ 
  baseURL: '...',
   onResponse: ...
 })
```

**Reply by Sin1tar** [reply]
> 

thank you so much!

**Reply by psycongaroo** [reply]
@pi0 
Thank you so much!
Is possible to implement in useFetch/useAsyncData the override of original $fetch?

So we can create an new instance and pass it like that?
https://pastebin.com/iJpRAy5z

**Reply by k0mar12** [reply]
@shinguweb96 I guess, we need to wait for this feature :)

**Reply by psycongaroo** [reply]
> @shinguweb96 I guess, we need to wait for this feature :)

I think too, but it will be so usefull 👍 
We can add interceptors with using useFetch / useAsyncData without loosing pick keys and all benefits of existent composables.

**Reply by psycongaroo** [reply]
@pi0 @danielroe @Atinux

**Reply by pi0** [reply]
@shinguweb96  Your workaround with `useCustomLazyFetch` seems a good idea. We might introduce a similar built-in like `createUseFetch` or `createFetch`  to provide default options (lazy can be an option too)

**Reply by psycongaroo** [reply]
> @shinguweb96 Your workaround with `useCustomLazyFetch` seems a good idea. We might introduce a similar built-in like `createUseFetch` or `createFetch` to provide default options (lazy can be an option too)

Thank you so much! 
Hope too see it soon :) 

And happy holidays to all team.

**Reply by ghost** [reply]
> ```js
> export const apiFetch = $fetch.create({ 
>   baseURL: '...',
>    onResponse: ...
>  })
> ```

@pi0 when trying to use this approach I get an error saying 
`Uncaught ReferenceError: $fetch is not defined at api.ts:1:25`
Is this a problem `$fetch.create()` not being supported anymore?

**Comment by aligzl** [comment]
what is final solution? @shinguweb96 your link not works

**Reply by k0mar12** [reply]
I have created folder - **tools**, dir for every manual importing plugins. With manual importing you can control order of importing and probably you wish include your fetch methods firstfull.

> /tools/spa-fetch.js

```
export default defineNuxtPlugin(({ ssrContext }) => {
    const { server } = useRuntimeConfig();

    return {
        provide: {
            spaFetch: $fetch.create({
                baseURL: server.baseURL,
                headers: useRequestHeaders(['cookie']),
                async onResponse({ response }) {
                    for (const cookie of parse(splitCookiesString(response.headers.get('set-cookie')))) {
                        setCookie(ssrContext.res, cookie.name, cookie.value, cookie);
                    }
                }
            })
        }
    }
});
```

> nuxt.config.ts

```
import { defineNuxtConfig } from 'nuxt3'

// https://v3.nuxtjs.org/docs/directory-structure/nuxt.config
export default defineNuxtConfig({
    publicRuntimeConfig: {
        server: {
            baseURL: process.env.APP_SERVER_URL
        }
    },

    plugins: [
        { src: '~/tools/spa-fetch.js' },
    ]
})
```

And then u can get **$spaFetch** in **nuxtApp**

In any plugin/component/pages etc and any of process env

P.S.:

But here is one problem. If you will create your fetch extension like composable feature, then u will get many instances, plugin - will create only one instance. So here can be memory leak, i guess.

**Comment by psycongaroo** [comment]
> @shinguweb96 Your workaround with useCustomLazyFetch seems a good idea. We might introduce a similar built-in like createUseFetch or createFetch to provide default options (lazy can be an option too)

@pi0 what about this?
